Reserving a drive/medium until locate expirancy is reached

SebaGougeaud opened the pull request.

SebaGougeaud committed:
lrs: add `locate_lock_expirancy` config parameter

This parameter forces one LRS to retain media if a locate was invoked
on them.

Signed-off-by: Sebastien Gougeaud <sebastien.gougeaud.externe@ayudata.fr>

SebaGougeaud committed:
db: add '3.2' schema

This schema alters the `lock` table to add a `last_locate` timestamp
that will be used to store the most recent locate invocation timestamp.

This timestamp is useful to determine is an LRS can release a medium or
not, i.e. if no recent locates requested this medium.

Signed-off-by: Sebastien Gougeaud <sebastien.gougeaud.externe@ayudata.fr>

SebaGougeaud committed:
types: add last_locate_timestamp field to lock

Signed-off-by: Sebastien Gougeaud <sebastien.gougeaud.externe@ayudata.fr>

SebaGougeaud committed:
dss: add last_locate lock field update on locate

Signed-off-by: Sebastien Gougeaud
<sebastien.gougeaud.externe@ayudata.fr>

SebaGougeaud committed:
dss: add behavior to `dss_lock_refresh`

This behavior allows one to only update the `last_locate` field on a
refresh operation, thus acknowledging a new locate command in the database.

Add a test ensuring the field is correctly updated.

Signed-off-by: Sebastien Gougeaud <sebastien.gougeaud.externe@ayudata.fr>

SebaGougeaud committed:
store: update locate timestamp in case lock is already taken

This behavior is useful to let the scheduler know when the last
locate was performed.

Tests are added to check this behavior.

Signed-off-by: Sebastien Gougeaud <sebastien.gougeaud.externe@ayudata.fr>

SebaGougeaud committed:
lrs: add last_locate check before picking a device

In case a device needs to be picked up, we check its medium is not
reserved for a future `get` operation before switching it to the
one requested by the client.

A test is added to check the daemon has the right behavior i.e.
fail a request if a drive/medium is reserved for a locate.

Signed-off-by: Sebastien Gougeaud <sebastien.gougeaud.externe@ayudata.fr>

YoannV-Ayudata commented on tests/externs/cli/test_locate.test:
```diff
@@ -158,6 +158,66 @@ function tape_cleanup
+
+function test_last_locate_timestamp
+{
+    sed -i "s/locate_lock_expirancy = 0/locate_lock_expirancy = 5000/g" $PHOBOS_CFG_FILE
```
major: restart the LRS with the proper environment variable instead, otherwise if the test fails, this is never changed back

YoannV-Ayudata commented on tests/externs/cli/test_locate.test:
```diff
@@ -158,6 +158,66 @@ function tape_cleanup
+    $phobos tape format "${tapes[1]}" &&
+        error "Format should have failed, as no drives are available"
+
+    sed -i "s/locate_lock_expirancy = 5000/locate_lock_expirancy = 0/g" $PHOBOS_CFG_FILE
```
suggest: add a check that after enough time has passed, the drive is available again. To make it easier, either change the date of the machine or use `faketime`

YoannV-Ayudata commented on tests/externs/cli/test_locate_common.sh:
```diff
@@ -231,6 +231,41 @@ function test_get_locate_cli
+    med=${med#\[\'}
+    med=${med%\'\]}
+
+    checkpoint1=$(date +%s)
+    sleep 1
+    $valg_phobos locate $obj
+    sleep 1
+    checkpoint2=$(date +%s)
+
+    db_time1=$($PSQL -t -c "SELECT last_locate FROM lock WHERE id = '${med}_legacy';")
+    db_time1=$(date -d "$db_time1" +%s)
+
+    if [ $checkpoint1 -gt $db_time1 ] || [ $checkpoint2 -lt $db_time1 ]; then
+        error "Invalid locate timestamp"
+    fi
```
suggest: since this particular check is already done in another test, you could instead do a simple locate here, get the time from the DB, sleep 1, locate again, and check the two times are different

YoannV-Ayudata commented on tests/interns/test_dss_lock.c:
```diff
@@ -457,6 +457,62 @@ static void dss_lock_hostname_unlock_ok(void **state)
+    assert_return_code(rc, -rc);
+
+    rc = dss_lock_status(handle, DSS_OBJECT, &GOOD_LOCKS[0], 1, &lock);
+    assert_int_equal(rc, -rc);
```
suggest: use `assert_return_code` like above instead

SebaGougeaud commented on src/cli/phobos/db/__init__.py:
```diff
@@ -726,6 +727,25 @@ def convert_2_2_to_3(self):
+        cur.execute(f"""
+            -- update lock table
+            ALTER TABLE lock ADD last_locate timestamp DEFAULT NULL;
+            UPDATE lock SET last_locate = NULL;
```
Done, you are right.

SebaGougeaud commented on src/lrs/lrs_sched.c:
```diff
@@ -1299,6 +1299,20 @@ static bool medium_is_write_compatible(struct media_info *medium,
     return true;
 }
 
+static bool medium_is_reserved_for_locate(struct media_info *medium,
```
Changed for `check_locate_expirancy`.

SebaGougeaud committed:
fix: test and doc improval

Signed-off-by: Sebastien Gougeaud <sebastien.gougeaud.externe@ayudata.fr>

SebaGougeaud commented on src/cli/phobos/db/sql/3.2/schema.sql:
```diff
@@ -0,0 +1,153 @@
+    hostname        varchar(256) NOT NULL,
+    owner           integer NOT NULL,
+    timestamp       timestamp DEFAULT now(),
+    is_early        boolean DEFAULT FALSE,
+    last_locate     timestamp DEFAULT NULL,
```
Yes, I suppose we can remove the boolean. I tried something.

SebaGougeaud committed:
types: remove `is_early` field of `pho_lock`

Also update the database scheme.

Signed-off-by: Sebastien Gougeaud <sebastien.gougeaud.externe@ayudata.fr>

patlucas commented on src/lrs/lrs_sched.c:
```diff
@@ -1299,6 +1299,17 @@ static bool medium_is_write_compatible(struct media_info *medium,
+    struct timespec expire;
+
+    expire.tv_sec = medium->lock.last_locate.tv_sec + lock_expirancy / 1000;
+    expire.tv_nsec = (lock_expirancy % 1000) * 1000000;
```
defect: (maybe I'm wrong) should we not also add medium->lock.last_locate.tv_nsec ? You have the add_timespec function that exists especially for that purpose. This function is already used with "is_past" in lrs_device.c or raid_common.c .

SebaGougeaud replied:
Done, I don't remember why, I thought the precision of a timestamp in the PSQL database was the second. Seems it's microsecond instead. Thanks.

patlucas commented on src/lrs/lrs_sched.c:
```diff
@@ -1364,6 +1379,16 @@ struct lrs_dev *dev_picker(GPtrArray *devices,
         }
 
+        if (lock_expirancy != 0) {
+            lrs_medium_update(&itr->ld_dss_media_info->rsc.id);
```
major: I do not understand why we check the lock here.

First, it is not mandatory that a device has an available loaded medium. (below, you will find example of testing ld_dss_media_info before calling function using it) .

Then, in my mind, lock expirancy is delay to prevent an other host to take a medium. Here, if a medium is loaded into a device, it means that this host already has the lock on this medium. So why checking any lock_expirancy protection ?

patlucas commented on tests/externs/cli/test_locate.test:
```diff
@@ -158,6 +159,73 @@ function tape_cleanup
+    sleep 1
+    checkpoint2=$(date +%s)
+
+    db_time=$($PSQL -t -c "SELECT last_locate FROM lock WHERE id = '${tape}_legacy';")
```
question: do we have any phobos function to list existing lock ? If yes, it is better to use them instead of using direct SQL requests.

SebaGougeaud replied:
We do not have a way to list Phobos locks.

```
phobos lock list
usage: phobos lock [-h] {clean} ...
phobos lock: error: argument verb: invalid choice: 'list' (choose from 'clean')
```

patlucas commented on tests/externs/cli/test_locate.test:
```diff
@@ -158,6 +159,73 @@ function tape_cleanup
+    invoke_lrs
+
+    $phobos locate $oid
+    $phobos tape format "${tapes[1]}" &&
```
comment: I understand here how this test works with your lock_expirancy test on dev_picker. But, I still do not understand the global lock_expirancy mechanism because in my mind it was just a way to prevent an other host to take this medium.

SebaGougeaud replied:
The use case was that the daemon released a medium which was located a few seconds ago. So we prevent a "kind of early" release, which prevents another host to take the given medium.

patlucas commented on tests/externs/cli/test_locate_common.sh:
```diff
@@ -231,6 +232,29 @@ function test_get_locate_cli
+    db_time2=$($PSQL -t -c "SELECT last_locate FROM lock WHERE id = '${med}_legacy';")
+    db_time2=$(date -d "$db_time2" +%s)
+
+    if [ $db_time1 -ge $db_time2 ]; then
```
question: is this test not redundant with the one you add in test_locate.test ?

SebaGougeaud replied:
Here we check the value of the `lock_expirancy` when the medium is not locked by the daemon. In `test_locate.test`, media are locked.

patlucas reviewed (changes requested):
I think I do not understand the global behavior of this lock_expirancy feature. Maybe is-it link with the existing "early" lock behavior that I need to understand first.

YoannV-Ayudata commented on src/core/dss/dss_lock.c:
```diff
@@ -109,6 +112,8 @@ static const char * const lock_query[] = {
                                " UPDATE lock SET timestamp = now()"
                                  WHERE_CONDITION
                                "END $$;",
+    [DSS_REFRESH_LOCATE_QUERY] = "UPDATE lock SET last_locate = now()"
```
Ok, thanks

YoannV-Ayudata commented on tests/externs/cli/test_locate.test:
```diff
@@ -198,24 +199,31 @@ function test_locate_update_timestamp_without_lock
         error "Format should have failed, as no drives are available"
 
-    sed -i "s/locate_lock_expirancy = 5000/locate_lock_expirancy = 0/g" $PHOBOS_CFG_FILE
+    date -s @"$(($(date +%s) + 60))"
```
defect: this is never changed back

SebaGougeaud committed:
test: LRS needs to go back from the future

Also fix a misalignment.

Signed-off-by: Sebastien Gougeaud <sebastien.gougeaud.externe@ayudata.fr>

SebaGougeaud commented on src/core/dss/dss_lock.c:
```diff
@@ -275,16 +283,20 @@ static int basic_lock(struct dss_handle *handle, enum dss_type lock_type,
+                        dss_type_names[lock_type], lock_id, lock_owner,
+                        lock_hostname);
+    else
+        g_string_printf(request, lock_query[DSS_REFRESH_LOCATE_QUERY],
```
`dss_lock_refresh()` was not used in the code prior to this patch, except for unit tests.

SebaGougeaud committed:
lrs: fix lock_expirancy checking

A check to the presence of a loaded medium was required. Also
take into account the number of microseconds of the database
timestamp.

Signed-off-by: Sebastien Gougeaud <sebastien.gougeaud.externe@ayudata.fr>

SebaGougeaud committed:
tests: setting back date if format fails

Signed-off-by: Sebastien Gougeaud <sebastien.gougeaud.externe@ayudata.fr>

SebaGougeaud committed:
tests: fix tape test issues

Also, correctly manage the medium lock retrieval through
`lrs_medium_update`.

Signed-off-by: Sebastien Gougeaud <sebastien.gougeaud.externe@ayudata.fr>

patlucas reviewed (approved):
I think I now understand the spirit of this patch. I think it does the job.

Beyond this patch, I think we need to upgrade the behaviour of lock and especially "locate locks" to remove them, especially when the expirancy is ended.

CourrierGui commented on src/lrs/lrs_cfg.c:
```diff
@@ -84,6 +84,11 @@ const struct pho_config_item cfg_lrs[] = {
+    [PHO_CFG_LRS_locate_lock_expirancy] = {
+        .section = "lrs",
+        .name    = "locate_lock_expirancy",
+        .value   = "300000",
```
A default value of 300s might be too much. It is probably best to disable this by default, no? Otherwise people will wonder why phobos is idle at the end of an I/O.

CourrierGui commented on tests/externs/cli/test_locate_common.sh:
```diff
@@ -231,6 +231,41 @@ function test_get_locate_cli
+    local obj="object_timestamp"
+
+    $phobos put /etc/hosts $obj || error "Error while putting $obj"
+    med=$($phobos extent list --output media_name $obj)
```
minor: add local declaration. Otherwise this variable will leak out of the function.

CourrierGui commented on src/lrs/lrs_sched.c:
```diff
@@ -1381,13 +1386,16 @@ struct lrs_dev *dev_picker(GPtrArray *devices,
-            lrs_medium_update(&itr->ld_dss_media_info->rsc.id);
-            if (check_locate_expirancy(itr->ld_dss_media_info,
-                                       lock_expirancy)) {
+            struct media_info *medium = lrs_medium_update(
```
suggest: this make a DSS request to fetch the new version of the medium. The device's medium keeps the old version. Since you update it here, you can also update the device's medium. To do so, you can look at what sched_medium_udpate does.

patlucas commented on tests/externs/cli/test_locate.test:
```diff
@@ -222,10 +222,12 @@ function test_last_locate_timestamp
-    $phobos tape format "${tapes[1]}" ||
-        error "Format should have succeeded"
+    $phobos tape format "${tapes[1]}"
+    local rc=$?
```
Done

patlucas commented on tests/externs/cli/test_locate.test:
```diff
@@ -216,18 +216,14 @@ function test_last_locate_timestamp
 
     invoke_lrs
 
+    date_start=$(date +%s)
```
Done
